src/Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.cpp

```diff
@@ -0,0 +1,111 @@
+#include <string_view>
+#include <Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/QueryNode.h>
+#include <Analyzer/SortNode.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class OptimizeRedundantFunctionsInOrderByVisitor : public InDepthQueryTreeVisitor<OptimizeRedundantFunctionsInOrderByVisitor>
+{
+    std::unordered_set<std::string_view> existing_keys;
+
+    bool isRedundantExpression(FunctionNode * function)
+    {
+        if (function->getArguments().getNodes().empty())
+            return false;
+        const auto & function_base = function->getFunction();
+        if (!function_base || !function_base->isDeterministicInScopeOfQuery())
+            return false;
+
+        // TODO: handle constants here
+        for (auto & arg : function->getArguments().getNodes())
+        {
+            switch (arg->getNodeType())
+            {
+                case QueryTreeNodeType::FUNCTION:
+                {
+                    if (!isRedundantExpression(arg->as<FunctionNode>()))
+                        return false;
+                    break;
+                }
+                case QueryTreeNodeType::COLUMN:
+                {
+                    auto * column = arg->as<ColumnNode>();
+                    if (!existing_keys.contains(column->getColumnName()))
```

Comment:
This should not be enough, because we can have multiple column nodes with same column name: `SELECT * FROM t1 INNER JOIN t2 ON t1.id = t2.id ORDER BY t1.id, t2.id.`. We need to compare columns using tree hash and `isEqual` function. There is a special utility file that contains hash tables for such use case.

Reply:
Yes, makes sense. I'll fix it.